crates/bevy_ecs/src/event.rs

```diff
@@ -852,4 +944,58 @@ mod tests {
         }
         read_for::<EmptyTestEvent>();
     }
+
+    #[test]
+    fn test_optional_event_reader_present() {
+        use bevy_ecs::prelude::*;
+
+        let mut world = World::new();
+        world.insert_resource(Events::<TestEvent>::default());
+
+        let mut reader = IntoSystem::into_system(|events: Option<EventReader<TestEvent>>| {
+            assert!(matches!(events, Some(_)));
+        });
+        reader.initialize(&mut world);
+        reader.run((), &mut world);
+    }
+
+    #[test]
+    fn test_optional_event_reader_not_present() {
+        use bevy_ecs::prelude::*;
+
+        let mut world = World::new();
+
+        let mut reader = IntoSystem::into_system(|events: Option<EventReader<TestEvent>>| {
+            assert!(matches!(events, None));
+        });
+        reader.initialize(&mut world);
+        reader.run((), &mut world);
+    }
+
+    #[test]
+    fn test_optional_event_writer_present() {
+        use bevy_ecs::prelude::*;
+
+        let mut world = World::new();
+        world.insert_resource(Events::<TestEvent>::default());
+
+        let mut reader = IntoSystem::into_system(|events: Option<EventWriter<TestEvent>>| {
+            assert!(matches!(events, Some(_)));
```

Comment:
I don't like the type elision here; it opens the door to really nasty bugs going undetected.